File: eoq_calculator.py
import numpy as np
import matplotlib.pyplot as plt

# Dữ liệu từ bài toán
D = 1040  # Nhu cầu hàng năm (20 chiếc/tuần × 52 tuần)
S = 60    # Chi phí đặt hàng mỗi lần (USD)
H = 25    # Chi phí lưu kho mỗi chiếc mỗi năm (25% × 100 USD)
Q = np.arange(50, 1000, 5)  # Phạm vi số lượng đặt hàng, bước nhảy 5 để chính xác hơn
print(Q);

# Tính toán chi phí
# ordering_cost = (D / Q) * S * np.ceil(D / Q)  # Chi phí đặt hàng
# holding_cost = (Q / 2) * H                    # Chi phí lưu kho
# total_cost = ordering_cost + holding_cost     # Tổng chi phí
ordering_cost = (D / Q) * S     # Chi phí đặt hàng (liên tục)
holding_cost = (Q / 2) * H      # Chi phí lưu kho
total_cost = ordering_cost + holding_cost  # Tổng chi phí

# Tính EOQ
EOQ = np.sqrt((2 * D * S) / H) # căn bậc 2 
min_total_cost = (D / EOQ) * S + (EOQ / 2) * H

# Tìm chỉ số gần nhất của EOQ trong mảng Q
closestQIndex = np.abs(Q - EOQ).argmin()
closestQ = Q[closestQIndex]
intersection_holding_cost = holding_cost[closestQIndex]
intersection_ordering_cost = ordering_cost[closestQIndex]

# Kiểm tra lý thuyết tại EOQ
print(f"Tại EOQ ≈ {EOQ:.2f} chiếc:")
print(f"Chi phí đặt hàng: {(D / EOQ) * S:.2f} USD")
print(f"Chi phí lưu kho: {(EOQ / 2) * H:.2f} USD")
print(f"Tổng chi phí: {min_total_cost:.2f} USD")
print("(Hai chi phí bằng nhau, khớp lý thuyết)")

# Vẽ biểu đồ
plt.figure(figsize=(12, 8))  # Kích thước lớn hơn để dễ đọc
plt.plot(Q, holding_cost, label='Chi phí lưu kho', color='red', linewidth=2)
plt.plot(Q, ordering_cost, label='Chi phí đặt hàng', color='blue', linewidth=2)
plt.plot(Q, total_cost, label='Tổng chi phí', color='green', linewidth=2)

# Đánh dấu điểm EOQ
plt.axvline(x=closestQ, color='purple', linestyle='--', label=f'EOQ ≈ {EOQ:.0f} chiếc (gần {closestQ})')
plt.scatter(closestQ, intersection_holding_cost, color='purple', s=100, label='Giao điểm chi phí lưu kho')
plt.scatter(closestQ, intersection_ordering_cost, color='purple', s=100, label='Giao điểm chi phí đặt hàng')
plt.axhline(y=min_total_cost, color='orange', linestyle='--', label=f'Tổng chi phí tối thiểu ≈ {min_total_cost:.0f} USD')

# Tùy chỉnh biểu đồ
plt.xlabel('Số lượng đặt hàng (chiếc)', fontsize=12)
plt.ylabel('Chi phí (USD)', fontsize=12)
plt.title('Mối quan hệ chi phí tồn kho và số lượng đặt hàng', fontsize=14, pad=15)
plt.legend(fontsize=10)
plt.grid(True, linestyle='--', alpha=0.7)
plt.tight_layout()

# Hiển thị biểu đồ
plt.show()

# Tính chi phí cụ thể
Q1, Q2 = 400, 500
total_cost_400 = (D / Q1 * S * np.ceil(D / Q1)) + (Q1 / 2) * H
total_cost_500 = (D / Q2 * S * np.ceil(D / Q2)) + (Q2 / 2) * H
print(f"Chi phí tồn kho với 400 chiếc: {total_cost_400:.0f} USD")
print(f"Chi phí tồn kho với 500 chiếc: {total_cost_500:.0f} USD")

File: eoq_calculator.js
const { ChartJSNodeCanvas } = require('chartjs-node-canvas');

// Dữ liệu từ bài toán
const D = 1040; // Nhu cầu hàng năm (20 chiếc/tuần × 52 tuần)
const S = 60;   // Chi phí đặt hàng mỗi lần (USD)
const H = 25;   // Chi phí lưu kho mỗi chiếc mỗi năm (25% × 100 USD)

// Tạo mảng số lượng đặt hàng (bước nhảy nhỏ hơn để mượt mà)
const Q = [];
for (let i = 50; i <= 1000; i += 5) Q.push(i); // Bước 5 để biểu đồ mượt hơn

// Tính toán chi phí (không ceil để khớp lý thuyết)
const orderingCost = Q.map(q => (D / q) * S); // Liên tục, không làm tròn
const holdingCost = Q.map(q => (q / 2) * H);
const totalCost = Q.map((q, index) => orderingCost[index] + holdingCost[index]);

// Tính EOQ
const EOQ = Math.sqrt((2 * D * S) / H);
const minTotalCost = (D / EOQ) * S + (EOQ / 2) * H;

// Tìm chỉ số gần nhất của EOQ
const closestQIndex = Q.reduce((closest, current, index) => {
  return Math.abs(current - EOQ) < Math.abs(Q[closest] - EOQ) ? index : closest;
}, 0);
const closestQ = Q[closestQIndex];
const intersectionHoldingCost = holdingCost[closestQIndex];
const intersectionOrderingCost = orderingCost[closestQIndex];

// Cấu hình và vẽ biểu đồ
const width = 1000; // Lớn hơn để dễ nhìn
const height = 700;
const chartCallback = (ChartJS) => {
  ChartJS.defaults.font.size = 16; // Font lớn hơn
  ChartJS.defaults.font.family = 'Arial'; // Font dễ đọc
};

const chartJSNodeCanvas = new ChartJSNodeCanvas({ width, height, chartCallback });

(async () => {
  const configuration = {
    type: 'line',
    data: {
      labels: Q,
      datasets: [
        {
          label: 'Chi phí lưu kho',
          data: holdingCost,
          borderColor: 'rgb(255, 0, 0)', // Đỏ rực
          backgroundColor: 'rgba(255, 0, 0, 0.1)',
          fill: true,
          tension: 0.3, // Làm đường cong mượt hơn
        },
        {
          label: 'Chi phí đặt hàng',
          data: orderingCost,
          borderColor: 'rgb(0, 0, 255)', // Xanh dương nổi
          backgroundColor: 'rgba(0, 0, 255, 0.1)',
          fill: true,
          tension: 0.3,
        },
        {
          label: 'Tổng chi phí',
          data: totalCost,
          borderColor: 'rgb(0, 128, 0)', // Xanh lá
          backgroundColor: 'rgba(0, 128, 0, 0.1)',
          fill: true,
          tension: 0.3,
        },
      ],
    },
    options: {
      scales: {
        x: { 
          title: { display: true, text: 'Số lượng đặt hàng (chiếc)', font: { size: 18 } },
          grid: { lineWidth: 0.5, color: 'rgba(0, 0, 0, 0.1)' } // Lưới mờ
        },
        y: { 
          title: { display: true, text: 'Chi phí (USD)', font: { size: 18 } }, 
          beginAtZero: true,
          grid: { lineWidth: 0.5, color: 'rgba(0, 0, 0, 0.1)' }
        },
      },
      plugins: {
        legend: { 
          position: 'top', 
          labels: { font: { size: 14 }, boxWidth: 20, padding: 20 },
          display: true
        },
        annotation: {
          annotations: [
            {
              type: 'line',
              mode: 'vertical',
              scaleID: 'x',
              value: closestQ,
              borderColor: 'purple',
              borderWidth: 3, // Dày hơn để nổi bật
              label: { 
                content: `EOQ ≈ ${EOQ.toFixed(0)} chiếc (gần ${closestQ})`, 
                enabled: true, 
                position: 'top',
                backgroundColor: 'rgba(128, 0, 128, 0.8)', // Nền tím mờ
                font: { size: 14 }
              },
            },
            {
              type: 'point',
              xValue: closestQ,
              yValue: intersectionHoldingCost,
              backgroundColor: 'purple',
              radius: 8, // To hơn
              label: { 
                content: `Giao điểm lưu kho ≈ ${intersectionHoldingCost.toFixed(0)} USD`, 
                enabled: true, 
                position: 'right',
                backgroundColor: 'white',
                font: { size: 12 }
              },
            },
            {
              type: 'point',
              xValue: closestQ,
              yValue: intersectionOrderingCost,
              backgroundColor: 'purple',
              radius: 8,
              label: { 
                content: `Giao điểm đặt hàng ≈ ${intersectionOrderingCost.toFixed(0)} USD`, 
                enabled: true, 
                position: 'left',
                backgroundColor: 'white',
                font: { size: 12 }
              },
            },
            {
              type: 'line',
              mode: 'horizontal',
              scaleID: 'y',
              value: minTotalCost,
              borderColor: 'orange',
              borderWidth: 2,
              label: { 
                content: `Tổng chi phí tối thiểu ≈ ${minTotalCost.toFixed(0)} USD`, 
                enabled: true, 
                position: 'left',
                backgroundColor: 'rgba(255, 165, 0, 0.8)',
                font: { size: 14 }
              },
            },
          ],
        },
      },
      elements: {
        line: { borderWidth: 2 } // Độ dày đường
      },
      layout: { padding: 20 } // Khoảng cách viền
    },
  };

  const image = await chartJSNodeCanvas.renderToBuffer(configuration);
  require('fs').writeFileSync('eoq_chart.png', image);

  console.log(`Chi phí tồn kho với 400 chiếc: ${(D / 400 * S * Math.ceil(D / 400) + (400 / 2) * H).toFixed(0)} USD`);
  console.log(`Chi phí tồn kho với 500 chiếc: ${(D / 500 * S * Math.ceil(D / 500) + (500 / 2) * H).toFixed(0)} USD`);
})();